### PR TITLE
Jwoo/ios action re write

### PR DESCRIPTION
**Pull request by nesalang**

### CODE REVIEW

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/CustomProgressBarRenderer.mm [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+@implementation CustomProgressBarRenderer
+
+- (UIView *)DeserializeToCustomUIElement:(NSData* )json
+{
+        if(json) {
+            NSArray *data = [NSJSONSerialization JSONObjectWithData:json options:NSJSONReadingMutableLeaves error:nil];
+            NSDictionary<NSString *, NSNumber *> *myval = [data objectAtIndex:0];
+            UIProgressView *progressView;
+            progressView = [[UIProgressView alloc] initWithProgressViewStyle:UIProgressViewStyleBar];
+            NSNumber *red = myval[@"red"];
```
Comment (yeohan) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/CustomProgressBarRenderer.mm]: Looks good!

File: source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/CustomParsingForIOSTest.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+namespace AdaptiveCardsSharedModelUnitTest
+{
+    TEST_CLASS(CustomParsingForIOSTest)
+    {
+    public:
+        TEST_METHOD(CanGetCustomJsonPayload)
+        {
+            std::string testJsonString = 
+            "{\
```
Comment (yeohan) [inline source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/CustomParsingForIOSTest.cpp]: nit - you may use `R"JSON(<string literal>)JSON"` to skip escaping special character, if we compile with C++ 11. See http://en.cppreference.com/w/cpp/language/string_literal

File: source/shared/cpp/ObjectModel/Enums.h [hunk 3]
```diff
@@ -223,7 +224,8 @@ enum class CardElementType
     TextInput,
     TimeInput,
     ToggleInput,
-    Custom
+    Custom,
+    CustomIOSDelegate,
```
Comment (yeohan) [inline source/shared/cpp/ObjectModel/Enums.h]: I can see how it works with _CustomIOSDelegate_, but what's the difference with the type _Custom_? Do we have code example for it for iOS?

File: samples/Tests/CustomParsingTestUsingProgressBar.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+                    "horizontalAlignment": "center"
+                },
+                {
+                    "type": "TextBlock",
+                    "text": "**Progress Bar**",
+                    "horizontalAlignment": "center"
+                },
+                {
+                    "type": "CustomIOSDelegate",
+                    "payload": [
```
Comment (yeohan) [inline samples/Tests/CustomParsingTestUsingProgressBar.json]: Looks good. I see customer has freedom to populate any data to consume in its own custom renderer.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm [hunk 5]
```diff
@@ -16,6 +17,41 @@
 @implementation ACOAdaptiveCard
 {
     std::shared_ptr<AdaptiveCard> _adaptiveCard;
+    NSMutableArray *_inputs;
+}
+
+- (void) setInputs:(NSMutableArray *) inputs
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm]: >NSMutableArray * [](start = 20, length = 16)

You should take an NSArray* here and copy its contents. You should not share ownership of the array with the caller. If the caller modifies the array after this call completes, that will change your array as well. Similarly, and even worse, if you change the array (by having someone call appendInputs), that will change the caller's local copy of the array too.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm [hunk 6]
```diff
@@ -16,6 +17,41 @@
...
+{
+    _inputs = inputs;
+}
+
+- (void) appendInputs:(NSMutableArray *)inputs
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm]: >NSMutableArray * [](start = 23, length = 16)

NSArray* here too.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm [hunk 7]
```diff
@@ -16,6 +17,41 @@
...
+
+- (NSData *)inputs:(NSError **)error
+{
+    if(_inputs){
+        NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
+        for(id<ACRIBaseInputHandler> input in _inputs)
+        {
+            if([input validate:error] == NO)
+            {
+                NSLog(@"input validation failed %@", *error);
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm]: >NSLog(@"input validation failed %@", *error); [](start = 16, length = 45)

If multiple errors are encountered, the error returned to the caller will be whatever the last error was.

How is someone supposed to call this method? What if the method returns an NSData* and also an NSError*? Is that a success or a failure?

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm [hunk 8]
```diff
@@ -16,6 +17,41 @@
...
+            }
+            else
+            {
+                [input getInput:dictionary];
+            }
+        }
+
+        error = nil;
+
+        return [NSJSONSerialization dataWithJSONObject:dictionary options:NSJSONWritingPrettyPrinted error:error];
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm]: >error [](start = 107, length = 5)

If you want to pass nil for the error here, just pass nil.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOBaseActionElement.h [hunk 9]
```diff
@@ -9,6 +9,18 @@
...
+typedef NS_ENUM(NSInteger, ACRActionType) {
+    ACRShowCard = 1,
+    ACRSubmit,
+    ACROpenUrl,
+};
+
+@property ACRActionType type;
+
+- (NSString *)title;
+- (NSString *)id;
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOBaseActionElement.h]: >id [](start = 14, length = 2)

I wouldn't recommend calling this method `id`. It's not illegal, but that identifier already has a meaning in Objective-C.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm (hunk 7 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm]: we don't have to validate inputs anymore, i will remove them. thank you


---
In reply to: [172390450](https://github.com/Microsoft/AdaptiveCards/pull/1211#discussion_r172390450) [](ancestors = 172390450)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOBaseActionElement.h (hunk 9 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOBaseActionElement.h]: thanks


---
In reply to: [172390942](https://github.com/Microsoft/AdaptiveCards/pull/1211#discussion_r172390942) [](ancestors = 172390942)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm (hunk 5 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm]: thanks!

---
In reply to: [172390193](https://github.com/Microsoft/AdaptiveCards/pull/1211#discussion_r172390193) [](ancestors = 172390193)

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/CustomProgressBarRenderer.mm (hunk 1 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/CustomProgressBarRenderer.mm]: thanks

---
In reply to: [172225868](https://github.com/Microsoft/AdaptiveCards/pull/1211#discussion_r172225868) [](ancestors = 172225868)

File: samples/Tests/CustomParsingTestUsingProgressBar.json (hunk 4 above)
Comment (nesalang) [reply samples/Tests/CustomParsingTestUsingProgressBar.json]: that's is correct

---
In reply to: [172233766](https://github.com/Microsoft/AdaptiveCards/pull/1211#discussion_r172233766) [](ancestors = 172233766)

File: source/shared/cpp/ObjectModel/Enums.h (hunk 3 above)
Comment (nesalang) [reply source/shared/cpp/ObjectModel/Enums.h]: Custom is registering custom parser written in c++. 
Sorry, this is out dated PR. the enum has changed from CustomIOSDelegate to Unknown

---
In reply to: [172232942](https://github.com/Microsoft/AdaptiveCards/pull/1211#discussion_r172232942) [](ancestors = 172232942)

File: source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/CustomParsingForIOSTest.cpp (hunk 2 above)
Comment (nesalang) [reply source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/CustomParsingForIOSTest.cpp]: thank you.

---
In reply to: [172231603](https://github.com/Microsoft/AdaptiveCards/pull/1211#discussion_r172231603) [](ancestors = 172231603)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm (hunk 5 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOAdaptiveCard.mm]: that's right. thanks


---
In reply to: [172582709](https://github.com/Microsoft/AdaptiveCards/pull/1211#discussion_r172582709) [](ancestors = 172582709,172392306,172390193)

**Pull request merged**